### PR TITLE
Fix deadlock during stopping CacheManager

### PR DESCRIPTION
**Pull request by vjuranek**
https://issues.jboss.org/browse/ISPN-6869

actual fix implemented by @danberindei , but as the fix was not sent for pretty long time and bug is still there sending PR myself

### CODE REVIEW

**Comment by danberindei** [comment]
@vjuranek Anxious to see the CI results, I think I didn't open a PR because I saw some test failures on my machine...

**Comment by vjuranek** [comment]
@danberindei hm, tests passed on my machine, but it seems that there's really some issue and testsuite has hanged. Would it be possible to take thread dump?

**Comment by danberindei** [comment]
@vjuranek Yes, there is a "View thread dump" link on the build page. After the build was killed (manually or after a timeout) you can also see the thread dump in the build log.

**Comment by vjuranek** [comment]
@danberindei there was a deadlock then the same cache is started from cache started listener. I hope the fix is trivial (PR updated), but let's wait for CI.

**Comment by vjuranek** [comment]
IMHO these two failures are not related to this change (and they are random, cannot reproduce locally)

**Comment by slaskawi** [comment]
Closing, please reopen once rebased.

**Comment by gustavonalle** [comment]
Friendly reminder: first reopen and force push AFTER. The other way around does not work :smile:

**Comment by vjuranek** [comment]
Please don't close PRs unless you really want to close them completely. I (and lots of other people) haven't sufficient rights to re-open PR, so @slaskawi could you please re-open? I've already rebased locally, as per @gustavonalle advice I'll push it once re-opened. Thanks!

**Comment by vjuranek** [comment]
rebased (thanks @rvansa for re-opening)

**Comment by vjuranek** [comment]
test failures are not related to this PR. (and can someone please remove "Needs rebase" label? thanks)

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java [hunk 1]
```diff
@@ -141,20 +142,22 @@
...
 
-   @GuardedBy("this")
-   private boolean stopping;
+   private final Lock lifecycleLock = new ReentrantLock();
+   private volatile boolean stopping;
 
    /**
-    * Constructs and starts a default instance of the CacheManager, using configuration defaults.  See {@link org.infinispan.configuration.cache.Configuration Configuration}
-    * and {@link org.infinispan.configuration.global.GlobalConfiguration GlobalConfiguration} for details of these defaults.
+    * Constructs and starts a default instance of the CacheManager, using configuration defaults.  See {@link
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: You should either revert these changes or change them so the link tags are on one line (and remove the labels, since they just duplicate the class name).

Comment (vjuranek) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: done

**Comment by vjuranek** [comment]
comments incorporated and rebased

**Comment by vjuranek** [comment]
import after rebase fixed

**Comment by wburns** [comment]
Hrmm a lot of these test failures are related to shutting down a server [1]. @vjuranek can you confirm they aren't related or fix? Thanks.

Also @danberindei how severe of an issue is this? Does it make sense for this to make it to 9.0 or can it wait until 9.1?

[1] http://ci.infinispan.org/viewLog.html?buildId=53463&buildTypeId=bt9

File: core/src/main/java/org/infinispan/registry/impl/InternalCacheRegistryImpl.java [hunk 2]
```diff
@@ -5,14 +5,12 @@
 import java.util.Set;
 import java.util.concurrent.ConcurrentMap;
 
-import org.infinispan.Cache;
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/registry/impl/InternalCacheRegistryImpl.java]: Removing this import breaks the build.

**Comment by vjuranek** [comment]
@tristantarrant rebased nad fixed
@wburns most of the test are flaky tests. I checked those which are not and these are not related to this change. Only `ClusterTopologyManagerTest#testAbruptLeaveAfterGetStatus2` **could** be related, but I'm not able to judge as I cannot reproduce locally.

**Comment by galderz** [comment]
@vjuranek Needs rebasing again, sorry about that! :$

**Comment by vjuranek** [comment]
rebased

**Comment by danberindei** [comment]
This is mostly my code so I won't integrate it myself, feel free to add comments/questions or to integrate it guys!

**Comment by gustavonalle** [comment]
Waiting for CI

**Comment by gustavonalle** [comment]
Integrated, thanks @danberindei !
